tls/s2n_protocol_preferences.c

```diff
@@ -17,39 +17,89 @@
 #include "error/s2n_errno.h"
 #include "utils/s2n_safety.h"
 
-int s2n_blob_set_protocol_preferences(struct s2n_blob *application_protocols, const char *const *protocols, int protocol_count)
+S2N_RESULT s2n_protocol_preferences_write(struct s2n_stuffer *protocol_stuffer, const uint8_t *protocol, size_t protocol_len)
 {
-    struct s2n_stuffer protocol_stuffer = {0};
+    ENSURE_MUT(protocol_stuffer);
+    ENSURE_REF(protocol);
+
+    /**
+     *= https://tools.ietf.org/rfc/rfc7301#section-3.1
+     *# Empty strings
+     *# MUST NOT be included and byte strings MUST NOT be truncated.
+     */
+    ENSURE(protocol_len != 0, S2N_ERR_INVALID_APPLICATION_PROTOCOL);
+    ENSURE(protocol_len <= 255, S2N_ERR_INVALID_APPLICATION_PROTOCOL);
+
+    uint32_t extension_len = s2n_stuffer_data_available(protocol_stuffer) + /* len prefix */ 1 + protocol_len;
+    ENSURE(extension_len <= UINT16_MAX, S2N_ERR_INVALID_APPLICATION_PROTOCOL);
+
+    GUARD_AS_RESULT(s2n_stuffer_write_uint8(protocol_stuffer, protocol_len));
+    GUARD_AS_RESULT(s2n_stuffer_write_bytes(protocol_stuffer, protocol, protocol_len));
 
-    GUARD(s2n_free(application_protocols));
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_protocol_preferences_set(struct s2n_blob *application_protocols, const char *const *protocols, int protocol_count)
+{
+    ENSURE_MUT(application_protocols);
 
+    /* NULL value indicates no preference so free the previous blob */
     if (protocols == NULL || protocol_count == 0) {
-        /* NULL value indicates no preference, so nothing to do */
-        return 0;
```

Comment:
Not to my knowledge. It just does the checks before calling free.

https://github.com/awslabs/s2n/blob/fc3f6c468f90e1decd3359c9ecbfd8560fa5e4eb/tls/s2n_protocol_preferences.c#L24-L29

Reply:
The advantage of the new implementation is we don't free the previous one until we've completely validated the new input. The previous behavior freed the previous list regardless, which would lead to an empty preference.